File: src/main/java/no/nav/fo/veilarbregistrering/registrering/bruker/InaktivBrukerService.java
package no.nav.fo.veilarbregistrering.registrering.bruker;

import no.nav.fo.veilarbregistrering.bruker.Bruker;
import no.nav.fo.veilarbregistrering.metrics.InfluxMetricsService;
import no.nav.fo.veilarbregistrering.metrics.PrometheusMetricsService;
import no.nav.fo.veilarbregistrering.oppfolging.OppfolgingGateway;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.transaction.annotation.Transactional;

import static no.nav.fo.veilarbregistrering.metrics.Events.MANUELL_REAKTIVERING_EVENT;

public class InaktivBrukerService {

    private static final Logger LOG = LoggerFactory.getLogger(InaktivBrukerService.class);

    private final BrukerTilstandService brukerTilstandService;
    private final ReaktiveringRepository reaktiveringRepository;
    private final OppfolgingGateway oppfolgingGateway;
    private final PrometheusMetricsService prometheusMetricsService;

    public InaktivBrukerService(
            BrukerTilstandService brukerTilstandService,
            ReaktiveringRepository reaktiveringRepository,
            OppfolgingGateway oppfolgingGateway,
            PrometheusMetricsService prometheusMetricsService) {
        this.brukerTilstandService = brukerTilstandService;
        this.reaktiveringRepository = reaktiveringRepository;
        this.oppfolgingGateway = oppfolgingGateway;

        this.prometheusMetricsService = prometheusMetricsService;
    }

    @Transactional
    public void reaktiverBruker(Bruker bruker, boolean erVeileder) {
        BrukersTilstand brukersTilstand = brukerTilstandService.hentBrukersTilstand(bruker);
        if (!brukersTilstand.kanReaktiveres()) {
            throw new KanIkkeReaktiveresException("Bruker kan ikke reaktiveres.");
        }

        reaktiveringRepository.lagreReaktiveringForBruker(bruker.getAktorId());
        oppfolgingGateway.reaktiverBruker(bruker.getGjeldendeFoedselsnummer());

        LOG.info("Reaktivering av bruker med aktørId : {}", bruker.getAktorId());

        if (erVeileder) {
            prometheusMetricsService.registrer(MANUELL_REAKTIVERING_EVENT);
        }

        AlderMetrikker.rapporterAlder(prometheusMetricsService, bruker.getGjeldendeFoedselsnummer());
    }
}

File: src/main/java/no/nav/fo/veilarbregistrering/registrering/bruker/InaktivBrukerService.kt
package no.nav.fo.veilarbregistrering.registrering.bruker

import no.nav.fo.veilarbregistrering.bruker.Bruker
import no.nav.fo.veilarbregistrering.log.loggerFor
import no.nav.fo.veilarbregistrering.metrics.Events
import no.nav.fo.veilarbregistrering.metrics.PrometheusMetricsService
import no.nav.fo.veilarbregistrering.oppfolging.OppfolgingGateway
import no.nav.fo.veilarbregistrering.registrering.bruker.AlderMetrikker.rapporterAlder
import org.springframework.transaction.annotation.Transactional

open class InaktivBrukerService(
    private val brukerTilstandService: BrukerTilstandService,
    private val reaktiveringRepository: ReaktiveringRepository,
    private val oppfolgingGateway: OppfolgingGateway,
    private val prometheusMetricsService: PrometheusMetricsService
) {
    @Transactional
    open fun reaktiverBruker(bruker: Bruker, erVeileder: Boolean) {
        val brukersTilstand = brukerTilstandService.hentBrukersTilstand(bruker)
        if (!brukersTilstand.kanReaktiveres()) {
            throw KanIkkeReaktiveresException("Bruker kan ikke reaktiveres.")
        }
        reaktiveringRepository.lagreReaktiveringForBruker(bruker.aktorId)
        oppfolgingGateway.reaktiverBruker(bruker.gjeldendeFoedselsnummer)
        LOG.info("Reaktivering av bruker med aktørId : {}", bruker.aktorId)
        if (erVeileder) {
            prometheusMetricsService.registrer(Events.MANUELL_REAKTIVERING_EVENT)
        }
        rapporterAlder(prometheusMetricsService, bruker.gjeldendeFoedselsnummer)
    }

    companion object {
        private val LOG = loggerFor<InaktivBrukerService>()
    }
}